src/backend/distributed/README.md

```diff
@@ -0,0 +1,2445 @@
+
+The aim of this planner is to avoid relying on PostgreSQL's standard_planner() for planning, which performs unnecessary computations like cost estimation, irrelevant for distributed planning. Skipping the standard_planner has significant performance gains for OLTP workloads. By focusing on "shard-reachable" queries, the Fast Path Router Planner is able to bypass the need for more computationally expensive planning processes, thereby accelerating query execution.
+
+### Main C Functions Involved:
+
+- `FastPathRouterPlan()`: The primary function for creating the fast-path query plan.
+- `FastPathRouterQuery()`: Validates if a query is eligible for fast-path routing by checking its structure and the WHERE clause.
+
+With  set client_min_messages to debug4; you should see the following in the DEBUG messages: "DEBUG:  Distributed planning for a fast-path router query"
+
+```sql
+-- Fetches the count of users born in the same year, but only
+-- for a single country
+SELECT EXTRACT(YEAR FROM date_of_birth) as birth_year, COUNT(*)
+FROM users_table
+WHERE country_code = 'USA' AND user_id = 15
+GROUP BY birth_year
+HAVING COUNT(*) > 10;
+```
+
+```sql
+-- all INSERT commands are by definition fast path
+-- router queries in the sense that they do not
+-- need any information from Postgres' standard_planner()
+INSERT INTO orders_table (user_id, product_id, order_date, status)
+VALUES (42, 555, now(), 'NEW');
+```
+
+```sql
+-- UPDATE/DELETEs can also be qualified as fast path router
+-- queries
+UPDATE products_table SET price = price * 1.1 WHERE product_id = 555;
+```
+
+Fast path queries have another important characteristic named "deferredPruning."
+
+For regular queries, Citus does the shard pruning during the planning phase, meaning that the shard that the query touches are calculated during the planning phase. However, in an ideal world, the shard pruning should happen during the execution and, for a certain class of queries, we support that. In the code, that is denoted by "Job->deferredPruning" field.
+
+Given that fast path queries are performance critical, they can be planned with prepared statements. When this is done, "Job->deferredPruning" becomes "true". And, the meaning of that is Citus can support PREPARED statements as expected. The first 6 executions of the plan do distributed planning, the rest is cached similar to Postgres' plan caching, and the shard pruning is done during the execution phase. And, if you attach a debugger, you'd see that on the first 6th execution, the debugger will stop at distributed_planner() function, but on the rest, it will not. The shard pruning for the cached command will happen in CitusBeginScan() function.
```

Comment:
```suggestion
Given that fast path queries are performance critical, they can be planned with prepared statements. When this is done, "Job->deferredPruning" becomes "true". And, the meaning of that is Citus can support PREPARED statements as expected. The first 6 executions of the plan do distributed planning, the rest is cached similar to Postgres' plan caching, and the shard pruning is done during the execution phase. And, if you attach a debugger, you'd see that on the first 6 executions, the debugger will stop at distributed_planner() function, but on the rest, it will not. The shard pruning for the cached command will happen in CitusBeginScan() function.
```